component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -345,21 +349,19 @@ private void processFirstStepOnly(AuthenticatedUser authenticatedUser, Authentic
      * @param username     the Username
      * @param tenantDomain the TenantDomain
      * @throws SMSOTPException
-     * @throws AuthenticationFailedException
+     * @throws UserStoreException
      */
     private void updateMobileNumberForUsername(AuthenticationContext context, HttpServletRequest request,
                                                String username, String tenantDomain)
-            throws SMSOTPException, AuthenticationFailedException {
+            throws SMSOTPException, UserStoreException {
 
-        if (username != null && !context.isRetrying()) {
-            if (log.isDebugEnabled()) {
-                log.debug("Updating mobile number for user : " + username);
-            }
-            Map<String, String> attributes = new HashMap<>();
-            attributes.put(SMSOTPConstants.MOBILE_CLAIM, String.valueOf(context.getProperty(SMSOTPConstants.REQUESTED_USER_MOBILE)));
-            SMSOTPUtils.updateUserAttribute(MultitenantUtils.getTenantAwareUsername(username), attributes,
-                    tenantDomain);
+        if (log.isDebugEnabled()) {
+            log.debug("Updating mobile number for user : " + username);
         }
+        Map<String, String> attributes = new HashMap<>();
+        attributes.put(SMSOTPConstants.MOBILE_CLAIM, String.valueOf(context.getProperty(SMSOTPConstants.REQUESTED_USER_MOBILE)));
+        SMSOTPUtils.updateUserAttribute(MultitenantUtils.getTenantAwareUsername(username), attributes,
+                tenantDomain);
```

Comment:
is the tenantDomain passed here same as user's tenant domain?
Shall we verify the scenario in a SaaS application as well?

Reply:
Tested and fixed an issue